text/0000-union-initialization-and-drop.md

```diff
@@ -0,0 +1,365 @@
+
+## Potential pitfalls around `DerefMut`
+
+When adding auto-derefs on the left-hand side of an assignment, as we traverse
+the path, once we hit a `union`, we stop adding further auto-derefs.  So with
+`s: Struct` and `u: Union`, when encountering `s.u.f.x`, auto-deref *does*
+happen on `s`, but not on `s.u` or any of the later components.
+
+Notice that this relies crucially on the only field of `ManuallyDrop` being
+private!  If we could project directly through that field, no `DerefMut` would
+be needed to reproduce the problematic example from the "guide" section.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This makes working with unions involving types that may have drop glue slightly
+more verbose than today: One has to write `ManuallyDrop` more often than one may
+want to.
+
+The restriction placed on `DerefMut` is not fully backwards compatible: A type
+could implement `Copy + DerefMut` and actually rely on the deref coercion inside
+a union.  That seems very unlikely, but should be tested with a crater run.
+
+The initialization tracking rules are somewhat surprising, and one might prefer
+the compiler to just not track anything when it comes to unions.  After all, the
+compiler fundamentally cannot know what part of the union is properly
+initialized.  Unfortunately, not having any initialization tracking is not an
+option when non-`Copy` fields are involved: We have to decide if moving out of a
+union field is allowed.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Ruling out fields with drop glue does not, in fact, reduce the expressiveness of
+unions because one can use `ManuallyDrop<T>` to obtain a drop-glue-free version
+of `T`.  If anything, having the `ManuallyDrop` in the union definition should
+help to drive home the point that no automatic dropping is happening, ever.
+(Before this RFC, automatic dropping is happening when assigning to a union
+field but not when the union goes out of scope.  That seems to be the result of
+necessity, not of a coherent design.)
+
+An alternative approach to proceed with unions has been
+[previously proposed by @petrochenkov](https://github.com/petrochenkov/rfcs/blob/e5266bd105f592f7408b8592c5c3deaccba7f1ec/text/1444-union.md#initialization-state).
+That proposal replaces RFC 1444 and goes into a lot more points than this much more
+limited proposal.  In particular, it allows fields with drop glue.  However, it
+can be pretty hard for the programmer to predict when drop glue will be
+automatically invoked on assignment or not, because the initialization tracking
+(which this RFC adapts from @petrochenkov's proposal) can sometimes be a little
+surprising when looking at individual fields: Whether `u.f2 = ...;` drops
+depends on whether `u.f1` has been previously initialized.  @petrochenkov hence
+proposes a lint to warn people that unions with drop-glue fields are not always
```

Comment:
The lint was inherited from the original RFC rather than proposed by me :)

Reply:
Ah, sorry. Fixed :)